src/parse_instruction.py

```diff
@@ -237,6 +237,24 @@ class NaiveParsingArch(ArchAsmParsing):
     def normalize_instruction(self, instr: AsmInstruction) -> AsmInstruction:
         return instr
 
+@dataclass
+class UsedRegNames:
+    used_names: Dict[Register, Register] = field(default_factory=dict)
+
+    def input_to_internal(self, reg: Register, arch: ArchAsmParsing) -> Register:
+        if reg in arch.aliased_regs:    
+            if reg not in self.used_names.values():
+                self.used_names[arch.aliased_regs[reg]] = reg
+            return arch.aliased_regs[reg]
+        else:
+            return reg
```

Comment:
Maybe something like this instead? It's stricter: it disallows using different aliases to refer to the same internal register?

```suggestion
    def input_to_internal(self, reg: Register, arch: ArchAsmParsing) -> Register:
        internal_reg = arch.aliased_regs.get(reg, reg)
        existing_reg = self.used_names.get(internal_reg)
        if existing_reg is None:
            self.used_names[internal_reg] = reg
        elif existing_reg != reg:
            raise DecompFailure(
                f"Source uses multiple names for {internal_reg} ({existing_reg}, {reg})"
            )
        return internal_reg
```